### PR TITLE
feature(#72) Improve Logging: Fix Duplicate Entries, Implement High-…

### PR DESCRIPTION
**Pull request by Bchir**
- Added 
  - new categorized Logging 
    - Ev.ServiceBus.LoggingExtensions.MessageProcessing : For logs related to message processing
    - Ev.ServiceBus.LoggingExtensions.ServiceBusClientManagement : For logs related to creation of client / disposition of clients
    - Ev.ServiceBus.LoggingExtensions.ServiceBusEngine : For logs related to the initialization of the Host
    - Ev.ServiceBus.HealthChecks.LoggingExtensionsHealthChecks : For logs related to service bus health checks registration
  - New telemetryOptions
    - ActivitySource (Ev.ServiceBus) for message processing 
    - Ev.ServiceBus.Apm : Elastic Apm integration
- Modified 
  - Reduce number of log entries and duplicate exception logging 
  - Use of high performance logging mechanism

To use OpenTelemetry

```csharp
  builder.WithTracing(tracing =>
        {
            tracing
                .AddSource("Ev.ServiceBus")
       });
```

To use Elastic APM you can use the new method UseApm
```csharp
  services.AddServiceBus(settings => {
        settings.Enabled = true;
        settings.ReceiveMessages = true;
        settings.WithConnection("", new ServiceBusClientOptions());
    }).UseApm();
```

`[Breaking]`
The exception catched during IMessageHandler was changed to FailedToProcessMessageException
Original exception is stored in the inner exception if you are using IExceptionHandler use this to get original message

```csharp
public class CustomExceptionHandler : IExceptionHandler
{
    public Task HandleExceptionAsync(ProcessErrorEventArgs args)
    {
            var original = exceptionEvent.Exception is FailedToProcessMessageException wrapperException
                ? wrapperException.InnerException!
                : exceptionEvent.Exception!;
        return Task.CompletedTask;
    }
}
```

### CODE REVIEW

File: src/Ev.ServiceBus/Management/Factories/MessageSenderFactory.cs [hunk 1]
```diff
@@ -30,7 +30,6 @@ public class MessageSenderFactory
     public IMessageSender CreateSender(ClientOptions[] senderOptions)
     {
         var options = (IClientOptions)senderOptions.First();
-        _logger.LogInformation("[Ev.ServiceBus] Initialization of sender client '{ResourceId}': Start", options.ResourceId);
```
Comment (benspeth) [inline src/Ev.ServiceBus/Management/Factories/MessageSenderFactory.cs:33]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: src/Ev.ServiceBus/Management/Factories/MessageSenderFactory.cs [hunk 2]
```diff
@@ -41,10 +40,7 @@ public IMessageSender CreateSender(ClientOptions[] senderOptions)
         }
 
         if (_options.Value.Settings.Enabled == false)
-        {
-            _logger.LogInformation("[Ev.ServiceBus] Initialization of sender client '{ResourceId}': Client deactivated through configuration", options.ResourceId);
```
Comment (benspeth) [inline src/Ev.ServiceBus/Management/Factories/MessageSenderFactory.cs:45]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: src/Ev.ServiceBus/Management/Wrappers/ReceiverWrapper.cs [hunk 3]
```diff
@@ -30,37 +31,33 @@ public class ReceiverWrapper
...
+        _serviceBusClientManagementLogger = _provider.GetRequiredService<ILogger<LoggingExtensions.ServiceBusClientManagement>>();
+        _messageProcessingLogger = _provider.GetRequiredService<ILogger<LoggingExtensions.MessageProcessing>>();
     }
 
     internal string ResourceId => _composedOptions.ResourceId;
     private ServiceBusProcessor? ProcessorClient { get; set; }
 
     public async Task Initialize()
     {
-        _logger.LogInformation("[Ev.ServiceBus] Initialization of receiver client '{ResourceId}': Start", _composedOptions.ResourceId);
```
Comment (benspeth) [inline src/Ev.ServiceBus/Management/Wrappers/ReceiverWrapper.cs:41]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: src/Ev.ServiceBus/Management/Wrappers/ReceiverWrapper.cs [hunk 4]
```diff
@@ -30,37 +31,33 @@ public class ReceiverWrapper
...
         if (_parentOptions.Settings.Enabled == false)
-        {
-            _logger.LogInformation(
-                "[Ev.ServiceBus] Initialization of client '{ResourceId}': Client deactivated through configuration", _composedOptions.ResourceId);
```
Comment (benspeth) [inline src/Ev.ServiceBus/Management/Wrappers/ReceiverWrapper.cs:45]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: src/Ev.ServiceBus/ServiceBusEngine.cs [hunk 5]
```diff
@@ -12,38 +12,32 @@ namespace Ev.ServiceBus;
...
+        _serviceBusEngineLogger = serviceBusEngineLogger;
     }
 
     public async Task StartAll()
     {
-        _logger.LogInformation("[Ev.ServiceBus] Starting azure service bus clients");
-
-        if (_options.Value.Settings.Enabled == false)
-        {
-            _logger.LogInformation("[Ev.ServiceBus] Reception and dispatch of messages have been deactivated through configuration.");
```
Comment (benspeth) [inline src/Ev.ServiceBus/ServiceBusEngine.cs:41]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: src/Ev.ServiceBus/ServiceBusEngine.cs [hunk 6]
```diff
@@ -12,38 +12,32 @@ namespace Ev.ServiceBus;
...
-        }
-        if (_options.Value.Settings.Enabled && _options.Value.Settings.ReceiveMessages == false)
-        {
-            _logger.LogInformation("[Ev.ServiceBus] Reception of messages have been deactivated through configuration.");
```
Comment (benspeth) [inline src/Ev.ServiceBus/ServiceBusEngine.cs:45]: Please keep this log. 
it is only produced during initialization of the app, so the cost of having them is low.
And they are useful to know exactly which resource are initialized

File: docs/CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,33 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## 6.0.0
```
[the remark this answers is not in the document]
Comment (Bchir) [reply docs/CHANGELOG.md:7]: there is a breaking change that is why am putting 6.0.0

File: src/Ev.ServiceBus.HealthChecks/LoggingExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+
+namespace Ev.ServiceBus.HealthChecks;
+
+public static class LoggingExtensions
+{
+    public record HealthChecks();
+
+    private static readonly Action<ILogger, string, string, Exception?> LogAddingHealthCheck =
```
[the remark this answers is not in the document]
Comment (Bchir) [reply src/Ev.ServiceBus.HealthChecks/LoggingExtensions.cs:10]: method for what ?

**Comment by benspeth** [comment]
merged in another PR

**Pull request closed without merging**